feat: switch logging to be asynchronous

paulswartz opened the pull request:
This changes the backend of the Splunk longer to match more closely the
architecture of the built-in [Console
logger](https://github.com/elixir-lang/elixir/blob/master/lib/logger/lib/logger/backends/console.ex).

- If we have no background tasks when a message comes in, send it
immediately.
- If we do have a background task and have room in the queue, queue it up to
be sent later and return immediately.
- If we do not have room in the queue, then wait for the existing request to
finish, then send the entire queue.

This applies some backpressure if we're unable to send messages, which will
then cause the Logger backend itself to apply backpressure to the application.

paulswartz committed:
test: switch tests to use Bypass

We'll be changing how we make the HTTP requests, to use async requests. Using
Bypass will allow us to verify that the correct requests are being made,
while we change those requests.

ryan-mahoney commented:
@paulswartz let me know if you are looking for a reviewer -- looks interesting!

paulswartz commented:
@ryan-mahoney It's a big refactoring so the more 👀 the better!

paulswartz committed:
chore: add ASDF + formatter configuration

paulswartz committed:
feat: switch logging to be asynchronous

This changes the backend of the Splunk longer to match more closely the
architecture of the built-in [Console
logger](https://github.com/elixir-lang/elixir/blob/master/lib/logger/lib/logger/backends/console.ex).

- If we have no background tasks when a message comes in, send it
immediately.
- If we do have a background task and have room in the queue, queue it up to
be sent later and return immediately.
- If we do not have room in the queue, then wait for the existing request to
finish, then send the entire queue.

This applies some backpressure if we're unable to send messages, which will
then cause the Logger backend itself to apply backpressure to the application.

ryan-mahoney commented on lib/logger_splunk_backend.ex:
```diff
@@ -29,115 +34,248 @@ defmodule Logger.Backend.Splunk do
     {:ok, state}
   end
 
+  def handle_http_response(%HTTPoison.AsyncStatus{code: 200}, state) do
```
should `handle_http_response` be a `defp`?

ryan-mahoney commented on lib/logger_splunk_backend.ex:
```diff
@@ -29,115 +34,248 @@ defmodule Logger.Backend.Splunk do
   end
 
   @unix_epoch :calendar.datetime_to_gregorian_seconds({{1970, 1, 1}, {0, 0, 0}})
   def ts_to_unix({date, {h, m, s, ms}}) do
```
should `ts_to_unix` be `defp`?

ryan-mahoney commented on lib/logger_splunk_backend.ex:
```diff
@@ -29,115 +34,248 @@ defmodule Logger.Backend.Splunk do
-      buffer: [],
-      buffer_size: 0
-    }
+  def maybe_send(%{response: nil, buffer_size: bs} = state) when bs != 0 do
```
should `maybe_send` be `defp`?

ryan-mahoney commented on lib/logger_splunk_backend.ex:
```diff
@@ -29,115 +34,248 @@ defmodule Logger.Backend.Splunk do
+    HTTPoison.post(state.host, IO.iodata_to_binary(state.buffer), headers, opts)
+  end
+
+  def await_response(%{response: %{id: ref}} = state) do
```
should `await_response` be `defp`?

ryan-mahoney reviewed (commented):
Read through and ran the tests, seems good!

Noted a few small things I saw around being more clear about which functions are not part of the modules external API.

Also, were specs intentionally omitted, or is that something you will be adding? 

It might be interesting to have a test that shows that in a case where the `buffer_length` is 10 and there are 100 messages, that 10 external calls were made (if that is always exactly predictable?).

Also, was wondering if there is any risk of losing a ~buffer's worth of messages if this process crashes and if so, if there is any reasonable mitigation, or if it's inevitable during that unlikely event.

paulswartz committed:
fix: ensure timestamps are always encoded as full floats

Encoding in scientific notation (example: 1.6802e9) breaks the Splunk JSON
parser. Instead of using Jason (which doesn't have an option for this), we
encode the float ourselves and build the JSON object.

paulswartz commented:
> Noted a few small things I saw around being more clear about which functions are not part of the modules external API.

The only functions which are actually part of the external API are `handle_event/5` (for log messages and flushing) and `handle_call/3` (for configure). Those are both part of the `:gen_event` behavior, so I've re-tagged them as such. I've also made the other functions private (minus `format_message` which needs to be public for testing).

> Also, were specs intentionally omitted, or is that something you will be adding?

The public functions (which have `@impl :gen_event`) implicitly have specs from their behavior. The others are (now) private, where we don't necessarily add specs. I added one for `format_message`.

> It might be interesting to have a test that shows that in a case where the `buffer_length` is 10 and there are 100 messages, that 10 external calls were made (if that is always exactly predictable?).

It really depends on how fast Splunk is responding. If the buffer is 10 but Splunk is returning quickly, there might be 100 external calls. There's one test which tests a slower response, and I've added an assertion about the count of messages there.
 
> Also, was wondering if there is any risk of losing a ~buffer's worth of messages if this process crashes and if so, if there is any reasonable mitigation, or if it's inevitable during that unlikely event.

It's always been possible, which is part of why this refactoring is important. Previously, if Splunk was slow or the network had issues, the process would crash and lose the messages. This will at least print the messages to `stdio` if Splunk is unable to accept the messages (see the ERROR tests with `FakeIO`).

paulswartz committed:
feat: allow configuring the Splunk index

paulswartz committed:
fix: can be installed with a default name

paulswartz committed:
feat: allow accepting `:all` metadata

paulswartz committed:
perf: reverse the metadata keys when setting

Instead of reversing the fetched metadata each time it's processed.

paulswartz commented:
@ryan-mahoney @losvedir I'm feeling confident in this now: I've been running it locally and the only errors I'm seeing are when my computer goes to sleep and the network drops. Ready for a review!

ryan-mahoney reviewed (approved):
Looks good!

losvedir commented on lib/logger_splunk_backend.ex:
```diff
@@ -5,140 +5,304 @@ defmodule Logger.Backend.Splunk do
+    metadata =
+      case Keyword.fetch(opts, :metadata) do
+        {:ok, :all} -> :all
+        {:ok, metadata} when is_list(metadata) -> Enum.reverse(metadata)
```
Why does it need to be reversed?

paulswartz replied:
It's reversed here so that the actual metadata processing in https://github.com/mbta/logger_splunk_backend/pull/6/files#diff-422fe5b31c1b6f41cc85cb125846e02cR248-R255 doesn't need to do a reverse on each call.

losvedir commented on lib/logger_splunk_backend.ex:
```diff
@@ -5,140 +5,304 @@ defmodule Logger.Backend.Splunk do
+
+    %{
+      state
+      | buffer: [state.buffer, data],
```
Not sure it's worth the optimization, but would it be faster if we did `buffer: [data | state.buffer]` here instead, and then did an `Enum.reverse` when it came time to actually send it? For that matter, since the event actually has a timestamp in it anyway, which Splunk will use to order properly, maybe we wouldn't even need the reversal.

paulswartz replied:
This creates an IO list (https://www.bignerdranch.com/blog/elixir-and-io-lists-part-1-building-output-efficiently/) which then gets converted to a binary all at once here: https://github.com/mbta/logger_splunk_backend/pull/6/files#diff-422fe5b31c1b6f41cc85cb125846e02cR229

For example:
```elixir
iex(1)> iolist = [[[["1" | "2"], "3"], "4", ["5", "6"]], "7", "8"]
[[[["1" | "2"], "3"], "4", ["5", "6"]], "7", "8"]
iex(2)> IO.iodata_to_binary(iolist)
"12345678"
```

losvedir commented on lib/logger_splunk_backend.ex:
```diff
@@ -5,140 +5,304 @@ defmodule Logger.Backend.Splunk do
+
+      true ->
+        # we're in the middle of the request, but have reached our max buffer size
+        state = buffer_event(level, msg, ts, md, state)
```
Should we still be buffering the event here if we've reached the max size? Looking at the definition of `buffer_event/5`, I don't see where it would reject the event, so I think we will continue to grow the buffer?

paulswartz replied:
You're correct, it does grow the buffer. The difference is that we wait for the buffer to be processed by `await_response`, instead of returning. The Logger keeps track of how many unprocessed messages there are in the message queue, so this will let them pile up in the background while we handle the HTTP response messages. If too many messages pile up in the queue, Logger will switch to sync mode and apply backpressure to clients, and eventually start discarding messages entirely.

losvedir commented on lib/logger_splunk_backend.ex:
```diff
@@ -5,140 +5,304 @@ defmodule Logger.Backend.Splunk do
+      true ->
+        # we're in the middle of the request, but have reached our max buffer size
+        state = buffer_event(level, msg, ts, md, state)
+        {:ok, await_response(state)}
```
I'm a bit confused on the state machine here. If we're in the middle of a request and the buffer is full (slow network), I would think we'd want to do something like just drop any incoming log events (or maybe `IO.puts` them). If I'm understanding `await_response/1` correctly, it blocks until it gets its HTTP messages. Could this potentially cause a user of the library, who's trying to `Logger.info(...)` in this situation, be blocked on that call?

paulswartz replied:
Yes, that's correct. The way it works is Logger sends this process a message for each log message. It also measures the length of the message queue, a proxy for unprocessed messages. If the message queue gets too long, at first Logger switches to sync mode, which doesn't return from `Logger.info/1` et al until the message is processed. If there are still too many messages, Logger will switch into discard mode where messages aren't sent to this process at all until the message queue drops far enough. The way we hook into that is by not processing more log messages when we've reached our buffer limit, so the messages queue up as messages in this process' queue.

losvedir replied:
Oh, I see, interesting! I wasn't considering the interactions at the `Logger` level w.r.t. to async and discarding. That makes sense.

losvedir reviewed (approved):
👍

paulswartz committed:
feat: try once to resend a message if it fails

This allows us to log a message even if there's a blip in logging to Splunk.

paulswartz committed:
doc: update README, add CHANGELOG

paulswartz commented:
@ryan-mahoney @losvedir last call! I did add some additional functionality in https://github.com/mbta/logger_splunk_backend/pull/6/commits/2ddfccb2665b935da0c1dfc07c5f40f2430eec62 to retry (once) messages that fail, and updated the documentation (and bumped the version to reflect the breaking change) in https://github.com/mbta/logger_splunk_backend/pull/6/commits/4f0bb51d3085a392f16250220c277aef180e22b4
